File: tools/t_class_classifier/cpp/feature_extract.cpp
#include "schemas.hpp"
#include <vector>
#include <set>
#include <algorithm>
#include <cmath>

bool connectivity_edges_are_one_based(const RealizedClosureTrace& trace) {
    if (trace.connectivity_record.num_vertices <= 0 || trace.connectivity_record.edges.empty()) {
        return false;
    }

    bool has_endpoint = false;
    for (const auto& edge : trace.connectivity_record.edges) {
        if (edge.size() < 2) {
            continue;
        }
        has_endpoint = true;
        int u = edge[0];
        int v = edge[1];
        if (u == 0 || v == 0) {
            return false;
        }
        if (u < 1 || u > trace.connectivity_record.num_vertices || v < 1 || v > trace.connectivity_record.num_vertices) {
            return false;
        }
    }
    return has_endpoint;
}

void compute_graph_metrics(
    const RealizedClosureTrace& trace,
    int& V,
    int& E,
    int& C,
    int& loop_count,
    int& raw_edge_count,
    int& parallel_edge_count
) {
    V = 0;
    if (!trace.closure_adjacency.empty()) {
        V = static_cast<int>(trace.closure_adjacency.size());
    } else if (trace.connectivity_record.num_vertices > 0) {
        V = trace.connectivity_record.num_vertices;
    }

    std::vector<std::set<int>> graph(V);
    raw_edge_count = 0;
    parallel_edge_count = 0;

    // Adjacency matrix is authoritative when present.
    if (!trace.closure_adjacency.empty()) {
        for (int i = 0; i < static_cast<int>(trace.closure_adjacency.size()); ++i) {
            for (int j = i + 1; j < static_cast<int>(trace.closure_adjacency[i].size()); ++j) {
                if (trace.closure_adjacency[i][j] != 0) {
                    graph[i].insert(j);
                    graph[j].insert(i);
                    raw_edge_count += 1;
                }
            }
        }
    } else if (trace.connectivity_record.num_vertices > 0 && !trace.connectivity_record.edges.empty()) {
        bool one_based_edges = connectivity_edges_are_one_based(trace);
        std::set<std::pair<int, int>> seen_edges;
        for (const auto& edge : trace.connectivity_record.edges) {
            if (edge.size() >= 2) {
                int u = edge[0];
                int v = edge[1];
                raw_edge_count += 1;

                if (one_based_edges) {
                    u -= 1;
                    v -= 1;
                }
                if (u >= 0 && u < V && v >= 0 && v < V) {
                    auto key = std::minmax(u, v);
                    if (seen_edges.find(key) != seen_edges.end()) {
                        parallel_edge_count += 1;
                    } else {
                        seen_edges.insert(key);
                        if (u != v) {
                            graph[u].insert(v);
                            graph[v].insert(u);
                        }
                    }
                }
            }
        }
    }

    // DFS connected components
    std::vector<bool> visited(V, false);
    C = 0;

    for (int i = 0; i < V; ++i) {
        if (!visited[i]) {
            std::vector<int> stack = {i};
            while (!stack.empty()) {
                int curr = stack.back();
                stack.pop_back();
                if (!visited[curr]) {
                    visited[curr] = true;
                    for (int neighbor : graph[curr]) {
                        if (!visited[neighbor]) {
                            stack.push_back(neighbor);
                        }
                    }
                }
            }
            C += 1;
        }
    }

    // Count unique undirected edges
    E = 0;
    for (int i = 0; i < V; ++i) {
        E += static_cast<int>(graph[i].size());
    }
    E /= 2;

    // Loop count
    if (V > 0) {
        loop_count = E - V + C;
    } else {
        loop_count = 0;
    }
}

int estimate_braid_proxy(const RealizedClosureTrace& trace, int loop_count) {
    if (loop_count <= 1) {
        return 1;
    } else if (loop_count == 2) {
        return 2;
    } else {
        return loop_count;
    }
}

File: tools/t_class_classifier/python/feature_extract.py
from typing import Tuple
from schemas import RealizedClosureTrace

def _normalize_connectivity_edge_indexing(trace: RealizedClosureTrace) -> bool:
    if not trace.connectivity_record or not trace.connectivity_record.edges:
        return False

    V = trace.connectivity_record.num_vertices
    endpoints = [
        node
        for edge in trace.connectivity_record.edges
        if len(edge) >= 2
        for node in edge[:2]
    ]
    if not endpoints or V <= 0:
        return False
    if any(node == 0 for node in endpoints):
        return False
    return all(1 <= node <= V for node in endpoints)

def compute_graph_metrics(trace: RealizedClosureTrace) -> Tuple[int, int, int, int, int, int]:
    # 1. Determine number of vertices V
    V = 0
    adj = trace.closure_adjacency
    if adj and len(adj) > 0:
        V = len(adj)
    elif trace.connectivity_record:
        V = trace.connectivity_record.num_vertices

    # 2. Build adjacency list representation (undirected)
    graph = {i: set() for i in range(V)}
    raw_edge_count = 0
    unique_edge_count = 0
    parallel_edge_count = 0

    # Ingest from closure_adjacency matrix. When present, this is the
    # authoritative graph source; the connectivity record is treated as
    # supporting provenance rather than a second edge source.
    if adj and len(adj) > 0:
        for i in range(len(adj)):
            for j in range(i + 1, len(adj[i])):
                if adj[i][j] != 0:
                    graph[i].add(j)
                    graph[j].add(i)
                    raw_edge_count += 1
                    unique_edge_count += 1
    elif trace.connectivity_record and trace.connectivity_record.edges:
        one_based_edges = _normalize_connectivity_edge_indexing(trace)
        seen_edges = set()
        for edge in trace.connectivity_record.edges:
            if len(edge) >= 2:
                u, v = edge[0], edge[1]
                raw_edge_count += 1
                if one_based_edges:
                    u -= 1
                    v -= 1
                if 0 <= u < V and 0 <= v < V:
                    key = tuple(sorted((u, v)))
                    if key in seen_edges:
                        parallel_edge_count += 1
                    else:
                        seen_edges.add(key)
                        unique_edge_count += 1
                        if u != v:
                            graph[u].add(v)
                            graph[v].add(u)

    # 3. Calculate components count C using DFS
    visited = [False] * V
    C = 0
    
    def dfs(node: int):
        stack = [node]
        while stack:
            curr = stack.pop()
            if not visited[curr]:
                visited[curr] = True
                for neighbor in graph[curr]:
                    if not visited[neighbor]:
                        stack.append(neighbor)

    for i in range(V):
        if not visited[i]:
            dfs(i)
            C += 1

    # 4. Count unique undirected edges E
    E = 0
    for u in graph:
        E += len(graph[u])
    E = E // 2

    # 5. Compute loop count (Betti-1 number)
    if V > 0:
        loop_count = E - V + C
    else:
        loop_count = 0

    return V, E, C, loop_count, raw_edge_count, parallel_edge_count

def estimate_braid_proxy(trace: RealizedClosureTrace, loop_count: int) -> int:
    if loop_count <= 1:
        return 1
    elif loop_count == 2:
        return 2
    else:
        return loop_count
